mark Instrumentation Key APIs Obsolete

TimothyMothra opened the pull request:
#2560 

## Changes
- mark PublicAPIs using Instrumentation Key as Obsolete.


<img width="1267" alt="image" src="https://user-images.githubusercontent.com/28785781/170109096-cd419c0d-50ef-4aa5-be9b-b93b4f582f32.png">

TimothyMothra commented on BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs:
```diff
@@ -74,14 +74,17 @@ static TelemetryConfiguration()
         /// Initializes a new instance of the TelemetryConfiguration class.
         /// </summary>
         /// <param name="instrumentationKey">The instrumentation key this configuration instance will provide.</param>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Transition to using connection strings for data ingestion. https://aka.ms/MigrateToConnectionString")]
```
@cartersocha, need to review the messaging here

TimothyMothra replied:
Spoke to Carter offline, he recommended using this link:
https://azure.microsoft.com/en-us/updates/technical-support-for-instrumentation-key-based-global-ingestion-in-application-insights-will-end-on-31-march-2025/

reyang reviewed (approved):
LGTM.

cijothomas commented on NETCORE/src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs:
```diff
@@ -44,6 +44,7 @@ public static IApplicationBuilder UseApplicationInsightsExceptionTelemetry(this
         /// <param name="services">The <see cref="IServiceCollection"/> instance.</param>
         /// <param name="instrumentationKey">Instrumentation key to use for telemetry.</param>
         /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Transition to using connection strings for data ingestion. https://aka.ms/MigrateToConnectionString")]
```
for an user of this API, this message is not telling how to migrate. The link is for an overall migration approach.

cijothomas commented:
@TimothyMothra I think we should offer specific alternate for each method we are Obsoleting.
For eg: service.AddAI(ikey) -- obsolete this, and suggest user to use service.AddAI(new AiOptions(){connectionstring}) ? Or create a github issue, document the migration there, and have obsolete message link to it. The github issue can further link to the aka.ms link to overall migration.

TimothyMothra commented:
@cijothomas I did both.
Updated the Obsolete message to point to our suggested alternative.
Also updated the linked issue with a full summary of these changes and as well as the relevant links to public docs.
Please re-review when you have a moment.

cijothomas commented on NETCORE/src/Microsoft.ApplicationInsights.WorkerService/ApplicationInsightsExtensions.cs:
```diff
@@ -21,6 +21,7 @@ public static partial class ApplicationInsightsExtensions
         /// <param name="services">The <see cref="IServiceCollection"/> instance.</param>
         /// <param name="instrumentationKey">Instrumentation key to use for telemetry.</param>
         /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Use an AddApplicationInsightsTelemetryWorkerService() overload to set ApplicationInsightsServiceOptions.ConnectionString. See https://github.com/microsoft/ApplicationInsights-dotnet/issues/2560 for more details.")]
```
the message is not very clear to me.
There is simply no other overload which takes ConnectionString. So suggesting to use an overload without being very clear which overload we are talking about can be confusing.

TimothyMothra replied:
I reviewed all the messages and added additional details.

I reworded this specific message to say `Use an AddApplicationInsightsTelemetryWorkerService() overload to set ApplicationInsightsServiceOptions.ConnectionString via Action<ApplicationInsightsServiceOptions>`
